opm/simulators/wells/BlackoilWellModel_impl.hpp

```diff
@@ -538,38 +516,32 @@ namespace Opm {
     BlackoilWellModel<TypeTag>::
     initFromRestartFile(const RestartValue& restartValues)
     {
-        const auto& defunctWellNames = ebosSimulator_.vanguard().defunctWellNames();
-
         // The restart step value is used to identify wells present at the given
         // time step. Wells that are added at the same time step as RESTART is initiated
         // will not be present in a restart file. Use the previous time step to retrieve
         // wells that have information written to the restart file.
         const int report_step = std::max(eclState().getInitConfig().getRestartStep() - 1, 0);
         const auto& summaryState = ebosSimulator_.vanguard().summaryState();
 
-        WellsManager wellsmanager(eclState(),
-                                  schedule(),
-                                  summaryState,
-                                  report_step,
-                                  Opm::UgGridHelpers::numCells(grid()),
-                                  Opm::UgGridHelpers::globalCell(grid()),
-                                  Opm::UgGridHelpers::cartDims(grid()),
-                                  Opm::UgGridHelpers::dimensions(grid()),
-                                  Opm::UgGridHelpers::cell2Faces(grid()),
-                                  Opm::UgGridHelpers::beginFaceCentroids(grid()),
-                                  grid().comm().size() > 1,
-                                  defunctWellNames);
-
-        const Wells* wells = wellsmanager.c_wells();
+        // Make wells_ecl_ contain only this partition's non-shut wells.
+        {
+            const auto& defunct_well_names = ebosSimulator_.vanguard().defunctWellNames();
+            auto is_shut_or_defunct = [&defunct_well_names](const Well2& well) {
+                return (well.getStatus() == Well2::Status::SHUT) || (defunct_well_names.find(well.name()) != defunct_well_names.end());
+            };
+            auto w = schedule().getWells2(report_step);
+            w.erase(std::remove_if(w.begin(), w.end(), is_shut_or_defunct), w.end());
+            wells_ecl_.swap(w);
+        }
 
-        wells_ecl_ = schedule().getWells2(report_step);
+        initializeWellPerfData();
 
```

Comment:
Why not move the initializeWellPerfData to the constructor in WellInterface_impl.hpp

Reply:
I was not sure the grid would always be load-balanced at construction, so I call the function only just before it is needed.